articles/how-to-generate-url-and-text-qr-codes-using-node/index.md

```diff
@@ -0,0 +1,335 @@
+
+> QRcode-Generator (Root Directory)
+>
+> > node_modules (folder)
+> >
+> > views (folder)
+> > > index.ejs (file)
+> > > scan.ejs (file)
+> >
+> > index.js (file)
+
+
+Let's begin by setting up our project directory. Create a folder and name it "QRcode-Generator". Open the folder with Visual Studio Code. Once inside VS Code, open the terminal. You can quickly do this by using the `` Ctrl+` `` shortcut on a PC or `` Control + Shift + ` `` shortcut on a Mac. Create a `package.json` file quickly by running `npm init -y` on the terminal. 
+
+Once created, run the following command:
+
+```bash
+npm i qrcode express body-parser nodemon ejs
+```
+
+This installs the following packages to the project:
+
+- body-parser 
+- ejs (Embedded JavaScript templates)
+- express 
+- qrcode
+- nodemon
+
+Let us add some nodemon configuration in the "package.json" file by adding the "start" and "dev" inside the scripts as follows:
+
+```json
+{
+    "name": "QRCODE-URL-main",
+    "description": "",
+    "version": "1.0.0",
+    "main": "index.js",
+    "scripts": {
+        "start": "npm index.js",
+        "dev": "nodemon index.js",
+        "test": "echo \"Error: no test specified\" && exit 1"
+    },
+    "keywords": [],
+    "author": "",
+    "license": "ISC",
+    "dependencies": {
+        "body-parser": "^1.19.0",
+        "ejs": "^3.1.6",
+        "express": "^4.17.1",
+        "nodemon": "^2.0.7",
+        "qrcode": "^1.4.4"
+    }
+}
+```
+
+### Adding a starting point
+
+We will need a starting point for our application. Create a file called "index.js".
+
+Inside the "index.js" file, let us do the following in sequential order:
+
+- Import the required modules.
+
+```javascript
+// Importing the packages that we need
+
+const express = require("express");
+const app = express();
+const port = 5000;
+const bp = require("body-parser");
+const qr = require("qrcode");
+```
+
+- Use express package to set our template engine (view engine) and the body-parser middleware for parsing bodies from URL and JSON objects.
+
+```javascript
+// Using the ejs (Embedded JavaScript templates) as our template engine
+// and call the body parser  - middleware for parsing bodies from URL
+//                           - middleware for parsing json objects
+
+app.set("view engine", "ejs");
+app.use(bp.urlencoded({ extended: false }));
+app.use(bp.json());
+```
+
+- Create a listener to the root route (`/`) in the "index.ejs" file.
+
+```javascript
+// Simple routing to the index.ejs file
+app.get("/", (req, res) => {
+    res.render("index");
+});
+```
+
+- Set blank error input error
+
+```javascript
+// Blank input
+// Incase of blank in the index.ejs file, return error 
+// Error  - Empty Data!
+
+app.post("/scan", (req, res) => {
+    const url = req.body.url;
+
+    if (url.length === 0) res.send("Empty Data!");
+    qr.toDataURL(url, (err, src) => {
+        if (err) res.send("Error occured");
+        res.render("scan", { src });
+    });
+});
+```

```

Comment:
This snippet is not explained clearly. Please explain that the function `qr.toDataURL` and how it will have the `err` variable set only if there is an error.

Reply:
Please confirm if it is well explained